### PR TITLE
Screen refresh rate improve by using condition_variable

### PR DESCRIPTION
**Pull request by pietro68**
Resolves: #3869 

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [x] Each commit compiles and runs on my machine without known undesirable changes of behavior

### CODE REVIEW

**Comment by pietro68** [comment]
It is only the last two commits - I can not seem to base this PR on vst2-fix-metering!

**Comment by Paul-Licameli** [comment]
Please rebase the last two commits only directly onto the release-3.2.2 branch.

**Comment by pietro68** [comment]
> Please rebase the last two commits only directly onto the release-3.2.2 branch.

Will do. And of course there was a missing header <condition_variable> on gcc builds which Visual Studio did not care about.

**Comment by Paul-Licameli** [comment]
Can you confirm that this really appears to correct the performance problem?

File: src/effects/RealtimeEffectState.cpp [hunk 1]
```diff
@@ -69,9 +69,14 @@ class RealtimeEffectState::AccessState : public NonInterferingBase {
...
    }
    void WorkerWrite() {
+
+      std::unique_lock lk(mLockForCV);
+
       // Worker thread avoids memory allocation.
       mChannelToMain.Write(CounterAndOutputs{
          mState.mWorkerSettings.counter, mState.mOutputs.get() });
+
+      mCV.notify_one();
```
Comment (Paul-Licameli) [inline src/effects/RealtimeEffectState.cpp:79]: As it says here https://en.cppreference.com/w/cpp/thread/condition_variable/notify_one, it is not necessary to keep the mutex locked while signalling, and might even be worse for performancde.  It needs to be held for the duration of changes that make the condition the waiter waits on true.

So you might insert `{` before the declaration of `lk` and `}` before you call `noitfy_one`.

File: src/effects/RealtimeEffectState.cpp [hunk 2]
```diff
@@ -179,11 +181,19 @@ struct RealtimeEffectState::Access final : EffectSettingsAccess {
...
+   */
+
    const EffectSettings &Get() override {
       if (auto pState = mwState.lock()) {
          if (auto pAccessState = pState->GetAccessState()) {
             if (pAccessState->mState.mInitialized)
-               FlushAttempt(*pAccessState); // try once, ignore success
+            {
+               // try once
+               assert(pAccessState->mState.mInitialized);
```
[the remark this answers is not in the document]
Comment (pietro68) [reply src/effects/RealtimeEffectState.cpp:184]: yes, I had to remove it when I rebased on release-3.2.2

File: src/effects/RealtimeEffectState.cpp (hunk 1 above)
Comment (pietro68) [reply src/effects/RealtimeEffectState.cpp:77]: However, I based my choice on what this page https://embeddedartistry.com/blog/2022/01/10/remember-to-lock-around-all-stdcondition_variable-variables/ says - search for this text: "However, it seems that the [C](https://embeddedartistry.com/fieldmanual-terms/c/)++ Core Guidelines discussions on GitHub prefer [this general guideline](https://github.com/isocpp/CppCoreGuidelines/issues/925#issuecomment-302269850):"

**Comment by pietro68** [comment]
> Can you confirm that this really appears to correct the performance problem?

To me it looks like when I move the sliders while playing, the play line advances much smoother now, but still not 100% smooth. Anyway, it has improved. This on Windows, I still have to try it on OSx.

File: src/effects/RealtimeEffectState.cpp (hunk 1 above)
Comment (pietro68) [reply src/effects/RealtimeEffectState.cpp:77]: but of course I will give a try to your suggestion

**Comment by crsib** [comment]
macOS has a pretty decent profiler. Why guess?

**Comment by Paul-Licameli** [comment]
> macOS has a pretty decent profiler. Why guess?

Why guess, when you can deduce?

**Review by Paul-Licameli — APPROVED** [review]
The total change might be small enough to squash into just one commit now

**Pull request merged**